### PR TITLE
Validate handshake messages.

### PR DESCRIPTION
**Pull request by boaks**
Convert handshake message fromBytearray into fromReader and check, if
bytes are still available after decoding the handshake messages.
Remove NULLClientKeyExchange.
Check length of handshake message with header fragment length field.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java [hunk 1]
```diff
@@ -429,13 +429,12 @@ private void processClientHello(int epoch, long sequenceNo, int messageSeq, byte
 	}
 
 	private static byte[] newHandshakeMessage(HandshakeType type, int messageSeq, byte[] fragment) {
-		int length = 8 + 24 + 16 + 24 + 24 + fragment.length;
```
Comment (boaks) [inline scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java]: That was a nice spotted bug. Detected by the new introduced check, if handshake message has left unprocessed payload.

**Comment by boaks** [comment]
The most changes are just replacing `byte[]` by `DatagramReader`.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/HandshakeMessage.java [hunk 2]
```diff
@@ -201,25 +201,27 @@ public static HandshakeMessage fromByteArray(byte[] byteArray, HandshakeParamete
 		DatagramReader reader = new DatagramReader(byteArray);
 		HandshakeType type = HandshakeType.getTypeByCode(reader.read(MESSAGE_TYPE_BITS));
 		LOGGER.trace("Parsing HANDSHAKE message of type [{}]", type);
-
+		int left = byteArray.length -MESSAGE_HEADER_LENGTH_BYTES;
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/HandshakeMessage.java]: format issue ?

File: scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java [hunk 3]
```diff
@@ -429,13 +429,12 @@ private void processClientHello(int epoch, long sequenceNo, int messageSeq, byte
...
 		DatagramWriter writer = new DatagramWriter();
 		writer.write(type.getCode(), 8);
-		writer.write(length, 24);
+		writer.write(fragment.length, 24);
 		writer.write(messageSeq, 16);
 		writer.write(0, 24); // fragment offset is always 0
-		writer.write(length, 24);
+		writer.write(fragment.length, 24);
```
[the remark this answers is not in the document]
Comment (boaks) [reply scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java]: [Handshake Message](https://tools.ietf.org/html/rfc6347#section-4.2.2)

[Fragmentation](https://tools.ietf.org/html/rfc6347#section-4.2.3)

> An unfragmented message is
   a degenerate case with fragment_offset=0 and fragment_length=length.

So, though fragmentation is not used in this test, `fragment_length=length` and so both are set to `fragment.length`.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java [hunk 4]
```diff
@@ -1086,22 +1086,16 @@ private DTLSMessage handshakeMessageFromByteArray(byte[] decryptedMessage) throw
 		//  are encapsulated within one or more TLSPlaintext structures, which
 		//  are processed and transmitted as specified by the current active session state."
 		if (LOGGER.isTraceEnabled()) {
-			LOGGER.trace("Decrypting HANDSHAKE message ciphertext{}{}", StringUtil.lineSeparator(),
+			LOGGER.trace("Parsing HANDSHAKE message plaintext{}{}", StringUtil.lineSeparator(),
```
[the remark this answers is not in the document]
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java]: Removed ...

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/HandshakeMessage.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/HandshakeMessage.java]: Yes, I want to speed up the build ;-).

File: scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbernard31) [reply scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/ServerHandshakerTest.java]: I get it now. Thx !

You can merge this one ;)

**Pull request merged**